External/FEXCore/Source/Interface/HLE/Thunks/Thunks.cpp

```diff
@@ -156,6 +234,125 @@ namespace FEXCore {
             }
         }
 
+        static void IsLibLoaded(void* ArgsRV) {
+            struct ArgsRV_t {
+                const char *Name;
+                bool rv;     // Guest function to call when branching to original_callee
+            };
+
+            auto &[Name, rv] = *reinterpret_cast<ArgsRV_t*>(ArgsRV);
+
+            auto CTX = Thread->CTX;
+            auto That = reinterpret_cast<ThunkHandler_impl*>(CTX->ThunkHandler.get());
+
+            {
+                std::shared_lock lk(That->ThunksMutex);
+                rv = That->Libs.contains(Name);
+            }
+        }
+
+        static void HostTrampolineForGuestcall(void* ArgsRV) {
```

Comment:
In the unlikely case of requesting a trampoline to a GuestTarget for which there already was a trampoline with different GuestUnpacker registered, this will leak the GuestcalInfo instance, right? Wouldn't it be safe to discard the old GuestcallToHostTrampoline entry in that scenario?

Reply:
It will keep both of them.

This is be needed as the unpacker also does the marshaling, the same function can need different trampolines - eg imagine the same callback getting passed to libA, and libB, both of which need to generate trampolines. There's no guarantee that the marshaling done in libA is compatible with the one done in libB.

In another very unlikely scenario, a function is guaranteed to be callable with different function signature from the cdecl/pcs abi, as freeing the stack frame is responsibility of the caller, and an application could, in theory, call the same function entry with two different signatures, and expect it to work.

This does causes an issue as `Trampoline(Guest, Packer1) != Trampoline(Guest, Packer2)` even though `Guest == Guest`. I don't think there's much we can do there, though.